File: ios/Classes/JanalyticsPlugin.m
#import "JanalyticsPlugin.h"

@interface NSError (FlutterError)
@property(readonly, nonatomic) FlutterError *flutterError;
@end

@implementation NSError (FlutterError)
- (FlutterError *)flutterError {
    return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %d", (int)self.code]
                               message:self.domain
                               details:self.localizedDescription];
}
@end


@implementation JanalyticsPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"janalytics"
                                     binaryMessenger:[registrar messenger]];
    JanalyticsPlugin* instance = [[JanalyticsPlugin alloc] init];
    instance.channel = channel;
    [registrar addApplicationDelegate: instance];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"setup" isEqualToString:call.method]) {
        [self setup:call result:result];
    }else if([@"setDebugMode" isEqualToString:call.method]){
        [self setDebugMode:call result:result];
    }else if([@"initCrashHandler" isEqualToString:call.method]){
        [self initCrashHandler:call result:result];
    }else if([@"stopCrashHandler" isEqualToString:call.method]){
        [self stopCrashHandler:call result:result];
    }else if([@"onPageStart" isEqualToString:call.method]){
        [self onPageStart:call result:result];
    }else if([@"onPageEnd" isEqualToString:call.method]){
        [self onPageEnd:call result:result];
    }else if([@"onCountEvent" isEqualToString:call.method]){
        [self onCountEvent:call result:result];
    }else if([@"onCalculateEvent" isEqualToString:call.method]){
        [self onCalculateEvent:call result:result];
    }else if([@"onLoginEvent" isEqualToString:call.method]){
        [self onLoginEvent:call result:result];
    }else if([@"onRegisterEvent" isEqualToString:call.method]){
        [self onRegisterEvent:call result:result];
    }else if([@"onBrowseEvent" isEqualToString:call.method]){
        [self onBrowseEvent:call result:result];
    }else if([@"onPurchaseEvent" isEqualToString:call.method]){
        [self onPurchaseEvent:call result:result];
    }else if([@"setAnalyticsReportPeriod" isEqualToString:call.method]){
        [self setAnalyticsReportPeriod:call result:result];
    }else if([@"identifyAccount" isEqualToString:call.method]){
        [self identifyAccount:call result:result];
    }else if([@"detachAccount" isEqualToString:call.method]){
        [self detachAccount:call result:result];
    }else {
        result(FlutterMethodNotImplemented);
    }
}

- (void) setup:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *appKey = arguments[@"appKey"];
    NSString *channel = arguments[@"channel"];
    JANALYTICSLaunchConfig *config = [[JANALYTICSLaunchConfig alloc] init];
    config.appKey = appKey;
    config.channel = channel;
    [JANALYTICSService setupWithConfig:config];
}

-(void) setDebugMode:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    id debug = arguments[@"debug"];
    if (debug) {
        [JANALYTICSService setDebug:YES];
    } else {
        [JANALYTICSService setDebug:NO];
    }
}

-(void) initCrashHandler:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    [JANALYTICSService crashLogON];
}

-(void) stopCrashHandler:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    
}

-(void) onPageStart:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *pageName = arguments[@"pageName"];
    [JANALYTICSService startLogPageView:pageName];
}

-(void) onPageEnd:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *pageName = arguments[@"pageName"];
    [JANALYTICSService stopLogPageView:pageName];
}

-(void) onCountEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    id eventId = arguments[@"eventId"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSCountEvent * event = [[JANALYTICSCountEvent alloc] init];
    event.eventID = eventId;
    if(![extMap isKindOfClass:[NSNull class]]){
        event.extra = extMap;
    }
    [JANALYTICSService eventRecord:event];
}

-(void) onCalculateEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    id eventId = arguments[@"eventId"];
    id eventValue = arguments[@"eventValue"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSCalculateEvent * event = [[JANALYTICSCalculateEvent alloc] init];
    event.eventID = eventId;
    event.value = [eventValue doubleValue];
    if (![extMap isKindOfClass:[NSNull class]]) {
        event.extra = extMap;
    }
    
    [JANALYTICSService eventRecord:event];
}

-(void) onLoginEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *loginMethod = arguments[@"loginMethod"];
    id loginSuccess = arguments[@"loginSuccess"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSLoginEvent * event = [[JANALYTICSLoginEvent alloc] init];
    event.method = loginMethod;
    NSLog(@"loginSuccess   %@ ",loginSuccess);
    if (![loginSuccess isKindOfClass:[NSNull class]]) {
        NSLog(@"loginSuccess");
        event.success = [loginSuccess boolValue];
    }
    if(![extMap isKindOfClass:[NSNull class]]){
        event.extra = extMap;
    }
    
    [JANALYTICSService eventRecord:event];
}

-(void) onRegisterEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *registerMethod = arguments[@"registerMethod"];
    NSNumber *registerSuccess = arguments[@"registerSuccess"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSRegisterEvent * event = [[JANALYTICSRegisterEvent alloc] init];
    event.method = registerMethod;
    if (![registerSuccess isKindOfClass:[NSNull class]]) {
        event.success = [registerSuccess boolValue];
    }
    if (![extMap isKindOfClass:[NSNull class]]) {
        event.extra = extMap;
    }
    [JANALYTICSService eventRecord:event];
}

-(void) onBrowseEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *browseId = arguments[@"browseId"];
    NSString *browseName = arguments[@"browseName"];
    NSString *browseType = arguments[@"browseType"];
    NSNumber *browseDuration = arguments[@"browseDuration"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSBrowseEvent * event = [[JANALYTICSBrowseEvent alloc] init];
    event.name = browseName;
    event.type = browseType;
    event.contentID =browseId;
    if (![browseDuration isKindOfClass:[NSNull class]]) {
        event.duration = [browseDuration longValue];
    }
    if (![extMap isKindOfClass:[NSNull class]]) {
        event.extra = extMap;
    }
    [JANALYTICSService eventRecord:event];
}

-(void) onPurchaseEvent:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString *purchaseGoodsid = arguments[@"purchaseGoodsid"];
    NSString *purchaseGoodsName = arguments[@"purchaseGoodsName"];
    NSNumber *purchasePrice = arguments[@"purchasePrice"];
    NSNumber *purchaseSuccess = arguments[@"purchaseSuccess"];
    NSNumber *purchaseCurrency = arguments[@"purchaseCurrency"];
    NSNumber *purchaseGoodsCount = arguments[@"purchaseGoodsCount"];
    NSString *purchaseGoodsType = arguments[@"purchaseGoodsType"];
    NSDictionary *extMap = arguments[@"extMap"];
    JANALYTICSPurchaseEvent * event = [[JANALYTICSPurchaseEvent alloc] init];
    if (![purchaseSuccess isKindOfClass:[NSNull class]]) {
        event.success = [purchaseSuccess boolValue];
    }
    if (![purchasePrice isKindOfClass:[NSNull class]]) {
        event.price = [purchasePrice doubleValue];
    }
    
    event.goodsName = purchaseGoodsName;
    event.goodsType = purchaseGoodsType;
    if (![purchaseGoodsCount isKindOfClass:[NSNull class]]) {
        event.quantity = [purchaseGoodsCount intValue];
    }
    event.goodsID = purchaseGoodsid;
    if(![purchaseCurrency isKindOfClass:[NSNull class]]){
        if ([purchaseCurrency intValue] == 0) {
            event.currency = JANALYTICSCurrencyCNY;
        }else{
            event.currency = JANALYTICSCurrencyUSD;
        }
    }
    if (![extMap isKindOfClass:[NSNull class]]) {
        event.extra = extMap;
    }
    [JANALYTICSService eventRecord:event];
}

-(void) setAnalyticsReportPeriod:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSNumber *period = arguments[@"period"];
    if (![period isKindOfClass:[NSNull class]] ) {
        [JANALYTICSService setFrequency:[period intValue]];
    }
    
}

-(void)identifyAccount:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    NSDictionary *arguments = call.arguments;
    NSString  *accountID = arguments[@"accountID"];
    NSNumber  *creationTime = arguments[@"creationTime"];
    NSNumber  *sex = arguments[@"sex"];
    NSNumber  *paid = arguments[@"paid"];
    NSString  *name = arguments[@"name"];
    NSString  *birthdate = arguments[@"birthdate"];
    NSString  *phone = arguments[@"phone"];
    NSString  *email = arguments[@"email"];
    NSString  *weiboID = arguments[@"weiboID"];
    NSString  *wechatID = arguments[@"wechatID"];
    NSString  *qqID = arguments[@"qqID"];
    NSDictionary  *extMap = arguments[@"extMap"];
    JANALYTICSUserInfo * userinfo = [[JANALYTICSUserInfo alloc] init];
    userinfo.accountID = accountID;
    if (![creationTime isKindOfClass:[NSNull class]]) {
        userinfo.creationTime = [creationTime longValue];
    }
    if (![sex isKindOfClass:[NSNull class]]) {
        userinfo.sex = [sex intValue];
    }
    if (![paid isKindOfClass:[NSNull class]]) {
        userinfo.paid = [paid intValue];
    }
    if (![email isKindOfClass:[NSNull class]] ) {
        userinfo.email = email;
    }
    if (![name isKindOfClass:[NSNull class]]) {
        userinfo.name = name;
    }
    if (![phone isKindOfClass:[NSNull class]]) {
        userinfo.phone = phone;
    }
    if (![birthdate isKindOfClass:[NSNull class]]) {
        userinfo.birthdate = birthdate;
    }
    if (![weiboID isKindOfClass:[NSNull class]]) {
        userinfo.weiboID = weiboID;
    }
    if (![wechatID isKindOfClass:[NSNull class]]) {
        userinfo.wechatID = wechatID;
    }
    if (![qqID isKindOfClass:[NSNull class]]) {
        userinfo.qqID = qqID;
    }
    if (![extMap isKindOfClass:[NSNull class]]) {
        NSArray *keys = [extMap allKeys];
        if (keys!=nil) {
            for (int i = 0; i<keys.count; i++) {
                NSString *key = keys[i];
                id value = [extMap objectForKey:key];
                if ([value isKindOfClass:[NSNull class]]) {
                    value = @"";
                }
                [userinfo setExtraObject:value forKey:key];
            }
        }
        
    }

    [JANALYTICSService identifyAccount:userinfo with:^(NSInteger err, NSString *msg) {
        
        reslut(@{@"code":@(err),@"msg":(msg?:@"")});
        
    }];
    
    
}

-(void)detachAccount:(FlutterMethodCall*)call result:(FlutterResult)reslut{
    [JANALYTICSService detachAccount:^(NSInteger err, NSString *msg) {
        
        reslut(@{@"code":@(err),@"msg":(msg?:@"")});
     
    }];
}

@end

File: android/src/main/java/com/jiguang/janalytics/JanalyticsPlugin.java
package com.jiguang.janalytics;

import android.content.Context;
import android.nfc.Tag;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import cn.jiguang.analytics.android.api.Account;
import cn.jiguang.analytics.android.api.AccountCallback;
import cn.jiguang.analytics.android.api.BrowseEvent;
import cn.jiguang.analytics.android.api.CalculateEvent;
import cn.jiguang.analytics.android.api.CountEvent;
import cn.jiguang.analytics.android.api.LoginEvent;
import cn.jiguang.analytics.android.api.Currency;
import cn.jiguang.analytics.android.api.PurchaseEvent;
import cn.jiguang.analytics.android.api.RegisterEvent;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import cn.jiguang.analytics.android.api.JAnalyticsInterface;

/** JanalyticsPlugin */
public class JanalyticsPlugin implements MethodCallHandler {

    private Context context;
    private MethodChannel channel;
    private JanalyticsPlugin instance;
    private ArrayList<String> getRidCache;
    private boolean dartIsReady ;



    /** Plugin registration. */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "janalytics");
        channel.setMethodCallHandler(new JanalyticsPlugin(registrar,channel));
    }

    private JanalyticsPlugin(Registrar registrar, MethodChannel channel) {
        this.context = registrar.context();
        this.channel = channel;
//      this.callbackMap = new HashMap<>();
//      this.sequence = 0;
        this.getRidCache = new ArrayList<>();
        instance = this;
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        if (call.method.equals("setup")) {
            setup(call,result);
        } else if (call.method.equals("setDebugMode")){
            setDebugMode(call,result);
        } else if (call.method.equals("initCrashHandler")){
            initCrashHandler(call,result);
        } else if (call.method.equals("stopCrashHandler")){
            stopCrashHandler(call,result);
        } else if (call.method.equals("onPageStart")){
            onPageStart(call,result);
        } else if (call.method.equals("onPageEnd")){
            onPageEnd(call,result);
        } else if (call.method.equals("onCountEvent")){
            onCountEvent(call,result);
        } else if (call.method.equals("onCalculateEvent")){
            onCalculateEvent(call,result);
        } else if (call.method.equals("onLoginEvent")){
            onLoginEvent(call,result);
        }else if (call.method.equals("onRegisterEvent")){
            onRegisterEvent(call,result);
        }else if (call.method.equals("onBrowseEvent")){
            onBrowseEvent(call,result);
        }else if (call.method.equals("onPurchaseEvent")){
            onPurchaseEvent(call,result);
        }else if (call.method.equals("setAnalyticsReportPeriod")){
            setAnalyticsReportPeriod(call,result);
        }else if(call.method.equals("identifyAccount")){
            identifyAccount(call,result);
        }else if (call.method.equals("detachAccount")){
            detachAccount(call,result);
        }else {
            result.notImplemented();
        }
    }

    public void setup(MethodCall call,Result result){
        JAnalyticsInterface.init(context);     		// 初始化 JPush
        Object channelObj = call.argument("channel");
        if (channelObj != null){
            String channel = (String) channelObj;
            JAnalyticsInterface.setChannel(context,channel);
        }
        dartIsReady = true;
    }

    public void setDebugMode(MethodCall call,Result result){
        Object debugObj = call.argument("debug");
        if (debugObj == null){
            return;
        }
        boolean debug = (boolean) debugObj;
        JAnalyticsInterface.setDebugMode(debug);
    }

    public void initCrashHandler(MethodCall call,Result result){
        JAnalyticsInterface.initCrashHandler(context);
    }

    public void stopCrashHandler(MethodCall call,Result result){
        JAnalyticsInterface.stopCrashHandler(context);
    }

    public void onPageStart(MethodCall call,Result result){
        String pageName = call.argument("pageName");
        JAnalyticsInterface.onPageStart(context,pageName);
    }

    public void onPageEnd(MethodCall call,Result result){
        String pageName = call.argument("pageName");
        JAnalyticsInterface.onPageEnd(context,pageName);
    }

    public void onCountEvent(MethodCall call,Result result){
        String eventId = call.argument("eventId");
        Map<String,String> extMap = call.argument("extMap");
        CountEvent event = new CountEvent(eventId);
        if (extMap!=null){
            event.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,event);
    }

    public void onCalculateEvent(MethodCall call,Result result){
        String eventId = call.argument("eventId");
        Object eventValueObj = call.argument("eventValue");
        if (eventValueObj == null){
            return;
        }
        double eventValue = (double) eventValueObj;
        Map<String,String> extMap = call.argument("extMap");
        CalculateEvent calculateEvent = new CalculateEvent(eventId,eventValue);
        if (extMap != null){
            calculateEvent.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,calculateEvent);
    }

    public void onLoginEvent(MethodCall call,Result result){
        String loginMethod = call.argument("loginMethod");
        Object loginSuccessObj = call.argument("loginSuccess");
        if (loginSuccessObj == null){
            return;
        }
        boolean loginSuccess = (boolean) loginSuccessObj;
        Map<String,String> extMap = call.argument("extMap");
        LoginEvent loginEvent = new LoginEvent(loginMethod,loginSuccess);
        if (extMap != null){
            loginEvent.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,loginEvent);
    }

    public void onRegisterEvent(MethodCall call,Result result){
        String registerMethod = call.argument("registerMethod");
        Object registerSuccessObj = call.argument("registerSuccess");
        if (registerSuccessObj == null){
            return;
        }
        boolean registerSuccess = (boolean) registerSuccessObj;
        Map<String,String> extMap = call.argument("extMap");
        RegisterEvent registerEvent = new RegisterEvent(registerMethod,registerSuccess);
        if (extMap != null){
            registerEvent.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,registerEvent);
    }

    public void onBrowseEvent(MethodCall call,Result result){
        String browseId = call.argument("browseId");
        String browseName = call.argument("browseName");
        String browseType = call.argument("browseType");
        Object browseDurationObj = call.argument("browseDuration");
        if (browseDurationObj == null){
            return;
        }
        int browseDuration = (int) browseDurationObj;
        Map<String,String> extMap = call.argument("extMap");
        BrowseEvent browseEvent = new BrowseEvent(browseId,browseName,browseType,browseDuration);
        if (extMap != null){
            browseEvent.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,browseEvent);
    }

    public void onPurchaseEvent(MethodCall call,Result result){
        String purchaseGoodsid = call.argument("purchaseGoodsid");
        String purchaseGoodsName = call.argument("purchaseGoodsName");
        Object purchasePriceObj = call.argument("purchasePrice");
        Object purchaseSuccessObj = call.argument("purchaseSuccess");
        Object purchaseCurrencyObj = call.argument("purchaseCurrency");
        Object purchaseGoodsCountObj = call.argument("purchaseGoodsCount");
        if (purchasePriceObj == null || purchaseSuccessObj == null || purchaseCurrencyObj == null || purchaseGoodsCountObj == null){
            return;
        }
        double purchasePrice = (double) purchasePriceObj;
        boolean purchaseSuccess = (boolean) purchaseSuccessObj;
        int purchaseCurrency = (int) purchaseCurrencyObj;
        String purchaseGoodsType = call.argument("purchaseGoodsType");
        int purchaseGoodsCount = (int) purchaseGoodsCountObj;
        Map<String,String> extMap = call.argument("extMap");
        PurchaseEvent purchaseEvent = new PurchaseEvent(purchaseGoodsid,purchaseGoodsName,purchasePrice,purchaseSuccess, Currency.values()[purchaseCurrency],purchaseGoodsType,purchaseGoodsCount);
        if (extMap != null){
            purchaseEvent.addExtMap(extMap);
        }
        JAnalyticsInterface.onEvent(context,purchaseEvent);
    }

    public void setAnalyticsReportPeriod(MethodCall call,Result result){
        Object periodObj = call.argument("period");
        if (periodObj == null){
            return;
        }
        int period = (int) periodObj;
        JAnalyticsInterface.setAnalyticsReportPeriod(context,period);
    }

    public void identifyAccount(MethodCall call, final Result result){
        String accountID = call.argument("accountID");
        if (accountID == null){
            return;
        }
        Object creationTimeObj = call.argument("creationTime");
        Object sexObj = call.argument("sex");
        Object paidObj = call.argument("paid");
        String name = call.argument("name");
        String birthdate = call.argument("birthdate");
        String phone = call.argument("phone");
        String email = call.argument("email");
        String weiboID = call.argument("weiboID");
        String wechatID = call.argument("wechatID");
        String qqID = call.argument("qqID");
        Map<String,String> extMap = call.argument("extMap");

        Account account = new Account(accountID);
        if (creationTimeObj != null){
            long creationTime ;
            if (creationTimeObj instanceof Long){
                creationTime = (long) creationTimeObj;
            }else{
                creationTime = (int)creationTimeObj;
            }
            account.setCreationTime(creationTime);        //账户创建的时间戳
        }
        if (sexObj != null){
            int sex = (int) sexObj;
            account.setSex(sex);
        }
        if (paidObj != null){
            int paid = (int) paidObj;
            account.setPaid(paid);
        }
        if (name != null){
            account.setName(name);
        }
        if (birthdate != null){
            account.setBirthdate(birthdate);       //"19880920"是yyyyMMdd格式的字符串
        }
        if (phone != null){
            account.setPhone(phone);
        }
        if (email != null){
            account.setEmail(email);
        }
        if (wechatID != null){
            account.setWeiboId(weiboID);
        }
        if (wechatID != null){
            account.setWechatId(wechatID);
        }
        if (qqID != null){
            account.setQqId(qqID);
        }
        if (extMap != null){
            Set<String> keys = extMap.keySet();
            for (String key : keys) {
                account.setExtraAttr(key,extMap.get(key));
            }
        }
        Log.d("shikk","account:"+account.toString());
        JAnalyticsInterface.identifyAccount(context, account, new AccountCallback() {
            @Override
            public void callback(int i, String s) {
                Map<String,Object> res = new HashMap<>();
                res.put("code",i);
                res.put("msg",s);
                result.success(res);
            }
        });


    }

    public void detachAccount(MethodCall call, final Result result){
        JAnalyticsInterface.detachAccount(context, new AccountCallback() {
            @Override
            public void callback(int i, String s) {
                Map<String,Object> res = new HashMap<>();
                res.put("code",i);
                res.put("msg",s);
                result.success(res);
            }
        });
    }

}
